### ISSUE TITLE
RoBERTa is not well implemented for tokenizers with pad_token_id != 1

### ISSUE THREAD

**Issue by dtamayo-nlp** [huggingface/transformers#34528]
### Feature request

I would like to request that RoBERTa models correctly accept tokenizers with pad_token_id != 1. This problem is inherit from [`fairseq`](https://github.com/facebookresearch/fairseq/blob/ecbf110e1eb43861214b05fa001eff584954f65a/fairseq/utils.py#L266) code.

### Motivation

**Problem Definition**
The current implementation of [Roberta](https://github.com/huggingface/transformers/tree/main/src/transformers/models/roberta) in transformers considers that the tokenizer has pad_token_id = 1:
- In the [TF implementation](https://github.com/huggingface/transformers/blob/405b56269812056d9593869e22b7b264d806cb1e/src/transformers/models/roberta/modeling_tf_roberta.py#L76) the pad token is explicitly set to 1.
- In all the implementations, the class `create_position_ids_from_input_ids` indirectly considers the pad_token_id = 1 in all the classes. ([modeling_roberta.py](https://github.com/huggingface/transformers/blob/9f06fb05059a973048f5865e7e385c9db5d6daa4/src/transformers/models/roberta/modeling_roberta.py#L1682), [modeling_tf_roberta.py](https://github.com/huggingface/transformers/blob/405b56269812056d9593869e22b7b264d806cb1e/src/transformers/models/roberta/modeling_tf_roberta.py#L125) and [modeling_flax_roberta.py](https://github.com/huggingface/transformers/blob/9f06fb05059a973048f5865e7e385c9db5d6daa4/src/transformers/models/roberta/modeling_flax_roberta.py#L73)).

**Motivation**
We have pre-trained a RoBERTa with another tokenizer from scratch and need to slightly change the current implementation to work correctly. The changes are minimal in `create_position_ids_from_input_ids`, we just need to change the `+ padding_idx` terms for `+ 1` terms in the incremental positions. This change will not affect the original implementations of RoBERTa.

### Your contribution

I can submit a PR with the modifications if you agree with the incorporation of this change.

**Comment by Rocketknight1** [comment]
Yes, the RoBERTa code is quite idiosyncratic, and it inherits some strange shortcuts from fairseq. The `position_ids` interact with the padding token idx in a very strange way that only really makes sense when it's a fixed value.

I think we could accept this PR, but we'd need:

1) Equivalent modifications for the TF, FLAX and PT files
2) Regression tests to make sure nothing breaks!

cc @ArthurZucker @LysandreJik in case they have objections to changing older code - we might have to reject this PR because of the risk of damaging backward compatibility

**Comment by ArthurZucker** [comment]
TBH I don't even mind only fixing the PT path! 🤗

**Comment by jklinck** [comment]
How would this not change the original implementation of RoBERTa if you are altering a hard coded variable? Wouldn't all other users need to input their own padding token for use?

**Comment by dtamayo-nlp** [comment]
Hi, 

I apologise for not closing this issue. I ended up realizing that it was easier changing the tokenizer and the position of the embedding of my custom RoBERTa-based model than changing all the classes of the HuggingFace code, I was afraid of failed deployment of my model with older HuggingFace versions.

Answering to @jklinck, the padding token of a RoBERTa-based model is set in certain model/tokenizer configs, the original idea was not only allowing my padding token to be implemented, but also allowing the code to extract that information from the original configs, this information is easy to extract. The real problem was an inaccurate use of the pad index in the function `create_position_ids_from_input_ids`.

This issue can definitely be solved, I close it for now due to the lack of time. If anyone wants to join, feel free to give it a try!

**Comment by jklinck** [comment]
I think I may have a fix, could you possibly reopen this.

**Comment by jklinck** [comment]
@dtamayo-nlp could you close this?

I did some more research and it is not possible to change the padding token. If you look at the json file that is at the bottom of this comment you can see all of the values assigned to to each token that roberta has been trained on. If you were to assign a different value to the padding token other than 1 it would break the model. If you were to say assign `padding_idx = 0`, then the model would not know the difference between a padding token and the start of sequence token ( `<s>`, this is assigned the value of 0). The only possible way to assign a different value to padding_idx is if you assigned it a value greater than 50,264 (the last token in roberta's training, this is at the end of the json file linked below). 

For example, if you were to assign `padding_idx = 6`, then the model would not be able to tell the difference between a padding token and a comma (in the json file below a comma is assigned the value of 6). Thus, anytime a comma is present then the model would think it is a padding token. Ultimately, this would break the model. 

If you do want to change the padding token to a value above 50,264 then all you need to do is go to modeling_roberta_tf.py and change line 75 from `self.padding_idx = 1` to `self.padding_idx = config.pad_token_id`. I have also written a test file if you would like it. 

https://huggingface.co/FacebookAI/roberta-base/raw/main/vocab.json
